### PR TITLE
-Support for Source Codecs

### LINKED ISSUE #1532: Support generic parsers/codecs
**Is your feature request related to a problem? Please describe.**

Both the [S3 Source](https://github.com/opensearch-project/data-prepper/issues/251) and HTTP Source use similar concepts of codecs for parsing input data. The S3 Source currently makes these codecs available as plugins. So they can be extended for the S3 source. But, if another source wanted to use these plugins it would be unable to.

**Describe the solution you'd like**

Create a core concept in Data Prepper of source-based codecs or parsers. These should be generic enough to take any Java `InputStream` and produce events from them.

I propose that we based this concept on the S3 codec. It has a few advantages:

1. It uses an InputStream. This is advantageous for large inputs2. 
2. It has a `Consumer` for each event. This also allows the source using the codec to receive `Event` objects and decide independently of the Codec of the best way to handle these.
3. It is not connected to HTTP in anyway.

**Describe alternatives you've considered (Optional)**

Data Prepper can have a similar concept for output codecs/parsers. However, I see no reason to force these to be the same concept. (Implementors may choose to pair them together to avoid code duplication).

**Additional context**

S3 Codec interface:

https://github.com/opensearch-project/data-prepper/blob/37c8b09ae8b7b8ce233564f5da7b09d1d41953f2/data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/codec/Codec.java#L19-L27

HTTP Codec interface:

https://github.com/opensearch-project/data-prepper/blob/37c8b09ae8b7b8ce233564f5da7b09d1d41953f2/data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/codec/Codec.java#L16-L23

### PR DESCRIPTION
**Pull request by umayr-codes**
Signed-off-by: umairofficial <umairhusain1010@gmail.com>

### Description
- `InputCodec` interface added in `data-prepper-api` package
- JSONCodec moved from `s3-source` to `parse-json-processor`
- CSVCodec moved from `s3-source` to `csv-processor`
- NewlineCodec moved from `s3-source` to `newline-codecs`
- Review comments incorporated
 
### Issues Resolved
Resolves #1532 
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/csv-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/csv/CsvInputCodec.java [hunk 1]
```diff
@@ -61,22 +62,12 @@ private void parseBufferedReader(final BufferedReader reader, final Consumer<Rec
         }
 
         MappingIterator<Map<String, String>> parsingIterator = mapper.readerFor(Map.class).with(schema).readValues(reader);
-        boolean hasNextValue;
```
Comment (dlvenable) [inline data-prepper-plugins/csv-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/csv/CsvInputCodec.java:64]: @umayr-codes , You have reverted some changes in your most recent rebase. See this commit: https://github.com/opensearch-project/data-prepper/commit/bc75494167d6befe3778493749678a501afbe1a9

I believe you can resolve this issue by changing all the code from this line through the end of this file back. Basically, we should not see any diffs after line 41.

You may also need to check the imports above.

File: data-prepper-plugins/csv-processor/src/test/java/org/opensearch/dataprepper/plugins/codec/csv/CsvCodecTest.java [hunk 2]
```diff
@@ -282,7 +280,7 @@ void test_when_manualHeaderTooManyColumns_then_omitsExtraColumnsOnParsedEvents(f
 
     @ParameterizedTest
     @ValueSource(ints = {2, 10, 100})
-    void test_when_autoDetectHeaderWrongNumberColumnsAndJaggedRows_then_skipsRemainingRows(final int numberOfRows) throws IOException {
```
Comment (dlvenable) [inline data-prepper-plugins/csv-processor/src/test/java/org/opensearch/dataprepper/plugins/codec/csv/CsvCodecTest.java:285]: Similar to my comment above about https://github.com/opensearch-project/data-prepper/commit/bc75494167d6befe3778493749678a501afbe1a9, this PR currently is reverting changes in main that need to remain.

You should revert back the changes starting at this line. There should be no diffs from this point forward.

Also, please check the imports.

**Review by dlvenable — APPROVED** [review]

**Review by dlvenable — CHANGES_REQUESTED** [review]
Removing accidental approval

File: data-prepper-plugins/csv-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/csv/CsvInputCodec.java (hunk 1 above)
Comment (umayr-codes) [reply data-prepper-plugins/csv-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/csv/CsvInputCodec.java]: Its done David!

File: data-prepper-plugins/csv-processor/src/test/java/org/opensearch/dataprepper/plugins/codec/csv/CsvCodecTest.java (hunk 2 above)
Comment (umayr-codes) [reply data-prepper-plugins/csv-processor/src/test/java/org/opensearch/dataprepper/plugins/codec/csv/CsvCodecTest.java]: Done! Thanks for pointing out.

**Review by kkondaka — APPROVED** [review]
Looks good.

**Comment by ashoktelukuntla** [comment]
Looks good Umair

**Review by dlvenable — APPROVED** [review]
Thanks @umayr-codes ! We should be able to merge once all the checks pass.

**Pull request merged**